### DISCUSSION TITLE
Use __dir__ to populate autocomplete for dynamic module attributes

### DISCUSSION THREAD

**Discussion by alkasm** [microsoft/pylance-release#4009] [Enhancement]
If you have a custom module `__getattr__`, it would be great to have some way to tell pylance what attributes are available.

Motivating example: a common lazy-loading mechanism for expensive module imports is to hide the import in a module-level `__getattr__` in an `__init__.py`:

```python
# mymodule.py
import importlib
from types import ModuleType

def __getattr__(attr: str) -> ModuleType:
    return importlib.import_module(__package__ + "." + attr, __package__)
```

Now I can "export" the available attributes in three main ways: `__dir__`, `__all__`, and annotations. 

### `__dir__`

From the [python docs](https://docs.python.org/3/library/functions.html#dir):
> If the object has a method named __dir__(), this method will be called and must return the list of attributes. This allows objects that implement a custom __getattr__() or __getattribute__() function to customize the way [dir()](https://docs.python.org/3/library/functions.html#dir) reports their attributes.

Hence `__dir__` is often used on an object with a custom `__getattr__` to potentially list the available attribute names that can be dynamically looked up. So pretty semantic to the suggestion.

The Python REPL autocomplete works with `__dir__` as well. If I define a custom one for `mymodule`:

```python
def __dir__() -> tuple[str]:
    return ("name1", "name2")
```
 
Then tab-completing `mymodule.<tab>` in the interpreter shows `mymodule.name1` and `mymodule.name2`. Furthermore that tab auto-complete also _evaluates_ each of those attributes, so it hits the custom `__getattr__` for each name. This behavior is the ideal UX in my mind.

### `__all__`

Another option is `__all__`, although this is typically used to _restrict_ the names in scope rather than _expand_ them. Either way, if pyright could resolve the strings from `__all__` via the `__getattr__`, I suppose this would also work.

### annotations

Another approach I tried was dropping an annotation for each attribute on the module:

```python
name1: ModuleType
name2: ModuleType
```

From this, pylance will successfully suggest the names for autocomplete on `mymodule.`. However, since they are a plain `ModuleType` and not the specific module, I don't get much help from `mymodule.name1.`.

---

Edit: Also in this specific example, there is a [stdlib mechanism for lazy imports](https://docs.python.org/3/library/importlib.html#importlib.util.LazyLoader). This ends up being more or less the same experience as using annotations---the correct names show up, but the attribute does not get evaluated and hence `mymodule.name1.` only shows generic `ModuleType` properties. The REPL experience here is as good, maybe even better than `__dir__`. Either way, lazy imports are not the only reason for a custom `__getattr__`, so I think this is an aside from the main suggestion. (then again maybe it is one of the more common uses of a custom module `__getattr__`?)

**Comment by erictraut** [comment]
Pylance is built on pyright, which is a static analysis tool. The behavior you're describing is too dynamic for a static analyzer to understand. Keep in mind that pyright doesn't actually execute any code.

One workaround is to use the conditional `typing.TYPE_CHECKING` to split the static-analysis code path from the dynamic (lazy) code path.

```python
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    # Use non-lazy (static) mechanism
    import foo
else:
    # Use lazy import mechanism
    def __getattr__(attr: str) -> ModuleType:
        return importlib.import_module(__package__ + "." + attr, __package__)
```

**Reply by alkasm** [reply]
Hey Eric, thanks a lot for the comments! Understood that pyright wouldn't provide this functionality OOTB, but thought that perhaps pylance could do a bit more digging here.

`if TYPE_CHECKING` guard (or similarly a separate .pyi?) is a good idea for a workaround!

**Comment by alkasm** [comment]
FWIW I think there's two different features from the REPL that are mentioned, which we could split the initial post on:

1. `__dir__` to populate a list of dynamic attributes
2. `__getattr__` to evaluate the attributes listed

I feel like (2) is probably heavy-handed, but (1) seems not that far out there.
